Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoASA.yaml

```diff
@@ -0,0 +1,576 @@
+      let src_or_any = set_union(srcipaddr_has_any_prefix, ipaddr_has_any_prefix); 
+      let dst_or_any = set_union(dstipaddr_has_any_prefix, ipaddr_has_any_prefix);
+      let allLogs = CommonSecurityLog
+                | where not(disabled)
+                | where (isnull(starttime) or TimeGenerated >= starttime) and (isnull(endtime) or TimeGenerated <= endtime)
+                | where DeviceVendor == "Cisco" and DeviceProduct == "ASA"
+                | where (isnull(dstportnumber) or (DestinationPort == dstportnumber)) or Message has tostring(dstportnumber)
+                | extend temp_isSrcMatch=has_any_ipv4_prefix(SourceIP,src_or_any), 
+                                temp_isDstMatch=has_any_ipv4_prefix(DestinationIP,dst_or_any)
+                | extend ASimMatchingIpAddr = case(
+                                                array_length(src_or_any) == 0 and array_length(dst_or_any) == 0, "-", // match not requested
+                                                (temp_isSrcMatch and temp_isDstMatch), "Both", // has to be checked before the individual 
+                                                temp_isSrcMatch, "SrcIpAddr",
+                                                temp_isDstMatch, "DstIpAddr",
+                                                "No match"
+                                            )
+                | where ASimMatchingIpAddr != "No match"
+                | where ((array_length(dvcaction) == 0) or SimplifiedDeviceAction has_any (dvcaction))
+                | where ((eventresult == "*") or 
+                        (eventresult == "Success" and SimplifiedDeviceAction == "Allow") or 
+                        (eventresult == "Failure" and SimplifiedDeviceAction != "Allow"))
+                | where DeviceEventClassID in ("106001","106006","106015","106016","106021","106022","106010","106014","106018","106023","302013","302015","302014","302016","302020","302021","710002","710003","710004","710005","106007","106017","106100","106002","106012","106013","106020");
+      let parsedData = allLogs
+                | where isnotempty(SourceIP)
+                | project-rename AliasDeviceEventClassID = DeviceEventClassID,
+                                 AliasDeviceVersion = DeviceVersion,
+                                 AliasDeviceAction = DeviceAction,
+                                 NetworkRuleName = DeviceCustomString2,
+                                 SrcIpAddr = SourceIP,
+                                 SrcPortNumber = SourcePort,
+                                 DstIpAddr = DestinationIP, 
+                                 DstPortNumber = DestinationPort
+                | project-away Activity, AdditionalExtensions, ApplicationProtocol, Device*, Destination*, EndTime, ExternalID, File*, Flex*, IndicatorThreatType, Malicious*, Old*, Process*, ReceiptTime, ReceivedBytes, Remote*, Report*, Request*, Source*, SentBytes, StartTime, TenantId, Threat*;
+      let unparsedData = allLogs
```

Comment:
You need to add prefiltering on the Message field for all unparsed branches. Currently you filter only after parsing.

My suggestion is to include the parsing of IP/Ports in the unparsed branch and pre and post filter in the unparsed branch. There are two branches in which the parsing may be different (where you add an optional user after). If needed created a 3rd root (parsed/unparsed/special_unparsed or whatever).

Reply:
Pre and post filters added.